presto-spi/src/main/java/com/facebook/presto/spi/type/SqlTimeWithTimeZone.java

```diff
@@ -83,6 +117,18 @@ public boolean equals(Object obj)
     @Override
     public String toString()
     {
-        return Instant.ofEpochMilli(millisUtc).atZone(ZoneId.of(timeZoneKey.getId())).format(formatter);
+        // FIXME This is hack that we need to use as the timezone interpretation depends on date (not only on time)
+        // Additionally we cannot just grab current TZ offset from timezoneOffsetReferenceTimestampUtc and use it
+        // as our formatting library will fail to display expected TZ symbol.
+        long currentMillisOfDay =
+                LocalTime.from(
+                        Instant.ofEpochMilli(timezoneOffsetReferenceTimestampUtc)
+                               .atZone(ZoneId.of(TimeZoneKey.UTC_KEY.getId())))
+                        .toNanoOfDay() / 1_000_000L; // nanos to millis
+        long timeMillisUtcInCurrentDay = timezoneOffsetReferenceTimestampUtc - currentMillisOfDay + millisUtc;
+
+        Instant utcInstantInCurrentDay = Instant.ofEpochMilli(timeMillisUtcInCurrentDay);
+        ZonedDateTime zonedDateTime = utcInstantInCurrentDay.atZone(ZoneId.of(timeZoneKey.getId()));
+        return zonedDateTime.format(formatter);
```

Comment:
One more thing. While range of allowed values is > 24h (48h? infinite?), `toString()` output only within 24h. That means two different values (and different by 24h or more) will be indistinguishable. :/